### ISSUE TITLE
EventExecutorGroups and Channel Initialization

### ISSUE THREAD

**Issue by d-t-w** [netty/netty#1236]
Found in 4.0.0.CR1:

I configure a ChannelPipeline with several handlers, the last of which can take a variable amount of time to execute. With that in mind I have the final handler run in its own EventExecutorGroup.

I had expected that the earlier handlers in the pipeline would execute regardless of whether the EventExecutor pinned for the final handler in the channel was currently busy or not, but find that if the EventExecutor pinned for the final handler is busy, then the ChannelInitializer is delayed in its execution, and the entire channel waits for the EventExecutor to become available before it is initialized.

This is because when the channel is initialized, it will execute the addLast using the configured EventExecutor if there is one:

DefaultChannelPipeline.addLast():181 

``` java
executeOnEventLoop(newCtx, new Runnable() {
    @Override
    public void run() {
        synchronized (DefaultChannelPipeline.this) {
            checkDuplicateName(name);
            addLast0(name, newCtx);
        }
    }
});
```

In my case, I am using Netty with the Http Codec support, so my pipeline consists of:
1. A handler which listens for channelRegistered, placing the channel in a ChannelGroup
2. The normal HttpResponseDecoder/Encoder codec netty ChannelHandlers
3. My custom handler which can sometimes take some time to execute.

I had expected that regardless of whether the entire pool of available Executors in my DefaultEventExecutorGroup were currently consumed, incoming Http Requests would be placed in the ChannelGroup and requests would decoded and queued on my final handler.

This is important, because amongst other things, if I want to close the server I want to send a 503 to all currently connected clients, so I want connection and request parsing to be independent of processing.

In case I am not configuring my bootsrap / executor group correctly, here is an example which demonstrates the condition:

``` java
public class NettyTestServer {

    private final long base;

    public NettyTestServer() throws InterruptedException {
        start();
        base = currentTimeMillis();
    }

    private void start() throws InterruptedException {
        bootstrap().bind().sync();
        System.out.println("Server started");
    }

    private ServerBootstrap bootstrap() {
        InternalLoggerFactory.setDefaultFactory(new Slf4JLoggerFactory());

        final EventExecutorGroup executorGroup = new DefaultEventExecutorGroup(2);

        try {
            InetSocketAddress address = new InetSocketAddress(InetAddress.getByName("127.0.0.1"), 8080);

            ServerBootstrap bootstrap = new ServerBootstrap();
            bootstrap.group(new NioEventLoopGroup(2))
                    .channel(NioServerSocketChannel.class)
                    .localAddress(address)
                    .childHandler(new ChannelInitializer<SocketChannel>() {
                        @Override
                        public void initChannel(SocketChannel ch) throws Exception {
                            ChannelPipeline pipeline = ch.pipeline();
                            pipeline.addLast(new RegisterHandler());
                            pipeline.addLast(executorGroup, new WaitHandler());
                        }
                    });

            return bootstrap;
        } catch (UnknownHostException e) {
            throw new RuntimeException("Netty configured on Unknown Host: ", e);
        }
    }

    class RegisterHandler extends ChannelStateHandlerAdapter {

        @Override
        public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
            System.out.println(stamp() + ": Channel Registered");
            ctx.fireChannelRegistered();
        }

        @Override
        public void inboundBufferUpdated(ChannelHandlerContext ctx) throws Exception {
            System.out.println(stamp() + ": Inbound Buffer Updated");
            ctx.fireInboundBufferUpdated();
        }
    }

    class WaitHandler extends ChannelStateHandlerAdapter {

        @Override
        public void inboundBufferUpdated(ChannelHandlerContext ctx) throws Exception {
            System.out.println(stamp() + ": Waiting");
            Thread.sleep(30000);
            System.out.println(stamp() + ": Done");
            ctx.fireInboundBufferUpdated();
            ctx.close();
        }
    }

    public String stamp() {
        return format("%s:%s - ", currentTimeMillis() - base, Thread.currentThread().getName());
    }

    public static void main(String[] args) throws InterruptedException {
        new NettyTestServer();
    }
}
```

As this is configured with two threads for both NIO and custom executorGroup, the third request will delay for 30 seconds before RegisterHandler.channelRegistered is called.

Three calls to:

```
curl -v -X PUT --data-binary "test-data" --header "Transfer-Encoding:" "127.0.0.1:8080/"
```

Will produce output similar too:

```
Server started
3503:nioEventLoopGroup-2-2 - : Channel Registered
3508:nioEventLoopGroup-2-2 - : Inbound Buffer Updated
3510:defaultEventExecutorGroup-1-1 - : Waiting
4548:nioEventLoopGroup-2-1 - : Channel Registered
4548:nioEventLoopGroup-2-1 - : Inbound Buffer Updated
4549:defaultEventExecutorGroup-1-2 - : Waiting
33511:defaultEventExecutorGroup-1-1 - : Done
33512:nioEventLoopGroup-2-2 - : Channel Registered
33524:nioEventLoopGroup-2-2 - : Inbound Buffer Updated
33525:defaultEventExecutorGroup-1-1 - : Waiting
34550:defaultEventExecutorGroup-1-2 - : Done
63525:defaultEventExecutorGroup-1-1 - : Done
```

One side-effect, related to https://github.com/netty/netty/issues/1230 is that regardless of how many threads you have in your EventLoopGroup, as the next one is selected in a round-robin way and not following a pool-like semantic where available threads are preferred, you can get channelInitialization delayed waiting for a handler to execute when there are actually plenty of threads idling/free.

**Comment by Satyan1611** [comment]
Hi  @normanmaurer, is this fix available in 4.1.31? as I observed the same round robin behavior in 4.1.31 version.

**Comment by Satyan1611** [comment]
Hi  @normanmaurer @trustin,  as I understand from this thread, the channelpipeline blocking has been fixed but the default way of choosing the next thread from EventExecutorGroup is still the round robin way. Kindly let me know if there is any way to pick the idle thread from the EventExecutorGroup instead of the round robin way.
